src/solvers/refinement/string_constraint.h

```diff
@@ -0,0 +1,180 @@
+/*******************************************************************\
+
+Module: String constraints.
+        These are formulas talking about strings. We implemented two
+        forms of constraints: `string_constraintt` implements formulas
+        of the form $\forall univ_var \in [lb,ub[. premise => body$,
+	and not_contains_constraintt implements those of the form:
+        $\forall x in [lb,ub[. p(x) => \exists y in [lb,ub[.
+	s1[x+y] != s2[y]$.
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_SOLVERS_REFINEMENT_STRING_CONSTRAINT_H
+#define CPROVER_SOLVERS_REFINEMENT_STRING_CONSTRAINT_H
+
+#include <langapi/language_ui.h>
+#include <solvers/refinement/bv_refinement.h>
+#include <solvers/refinement/refined_string_type.h>
+
+class string_constraintt: public exprt
+{
+public:
+  // String constraints are of the form
+  // forall univ_var in [lower_bound,upper_bound[. premise => body
+  // or premise => body
+
+  inline const exprt &premise() const
+  {
+    return op0();
+  }
+
+  inline const exprt &body() const
+  {
+    return op1();
+  }
+
+  inline const symbol_exprt &univ_var() const
+  { return to_symbol_expr(op2()); }
+
+  inline const exprt &upper_bound() const
+  {
+    return op3();
+  }
+
+  inline const exprt &lower_bound() const
+  {
+    return operands()[4];
+  }
+
+
+ private:
+  string_constraintt();
+
+ public:
+  string_constraintt(
```

Comment:
Actually now, all object of `string_constraintt` are of the same kind, the other constructors are just with default values for some parameters.  Maybe one of the comment at the begging was confusing, I removed the erroneous part https://github.com/diffblue/cbmc/pull/374/commits/418d33b3c08e4257e48cb425ca7bc4203218a494
Note that this is part of another pull request, I will rebase this one when the other get merged.